source/features/useful-not-found-page.tsx

```diff
@@ -36,54 +38,55 @@ function parseCurrentURL(): string[] {
 	return parts;
 }
 
-// If the resource was deleted, link to the commit history
-async function addCommitHistoryLink(bar: Element): Promise<void> {
-	const parts = parseCurrentURL();
-	parts[2] = 'commits';
-	const url = '/' + parts.join('/');
-	if (await is404(location.origin + url)) {
+async function getLatestChangeToFile(): Promise<Record<string, any> | void> {
+	const {branch, filePath} = new GitHubURL(location.href);
+	if (!branch || !filePath) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			See also the {getType()}’s <a href={url}>commit history</a>.
-		</p>,
-	);
-}
-
-// If the resource exists in the default branch, link to it
-async function addDefaultBranchLink(bar: Element): Promise<void> {
-	const parts = getCleanPathname().split('/');
-	const branch = parts[3];
-	if (!branch) {
+	const {repository} = await api.v4(`
+		repository() {
+			object(expression: "${branch}") {
+				... on Commit {
+					history(first: 1, path: "${filePath}") {
+						nodes {
+							oid
+						}
+					}
+				}
+			}
+		}
+	`);
+	const [commit] = repository.object.history.nodes;
+	if (!commit) {
 		return;
 	}
 
-	const defaultBranch = await getDefaultBranch();
-	if (branch === defaultBranch) {
-		return;
+	// API v4 doesn't support retrieving a list of changed files for a commit:
+	// https://github.community/t/graphql-api-get-list-of-files-related-to-commit/14047/2
+	const commitInfo = await api.v3(`commits/${commit.oid as string}`);
```

Comment:
We could, but we require some info (commit date and time, url to last known version of file, parent sha in case of rename, url to view commit) from the `commitInfo` object as well.

Reply:
If you are talking about the types, you can add what ever types are missing to it.

Reply:
Sorry, I didn't understand. Should I replace the API call and comment I'm using with the function, and modify the function to return the `commitInfo` as well?

Reply:
> Sorry, I didn't understand. Should I replace the API call and comment I'm using with the function

Yes



> and modify the function to return the `commitInfo` as well?

Not sure I am understanding this comment.

The API is returning the same as you are here, its the same call. 
If you need more info from the type, just add it to the interface there.

Reply:
I think we could rename the current function to `getLatestCommitForFile` and return only `commitInfo`. The caller can parse the object to get the exact change to the file (`fileInfo`), by calling another function `getChangeToFileInCommit`. Does that sound ok?

Reply:
I am lost here.

All I am asking it to rename and export

https://github.com/sindresorhus/refined-github/blob/68abae5ea744ff82ea9682c7d9cefc1949805aee/source/features/follow-file-renames.tsx#L16-L20

And replace
https://github.com/gamemaker1/refined-github/blob/cf6ab76a30be695a43681cfc9936b8d50dd544ca/source/features/useful-not-found-page.tsx#L65-L67

With `const commitInfo = await whatEverYouNamedTheFunction()`

Reply:
Sorry, let me try explaining it more clearly:

To display the improved 404 page, we need the following data:
	R1. list of files that was changed by the commit
	R2. datetime of the commit
	R3. URL to the commit page
	R4. parent commit sha

The function `findRename` from `follow-file-renames.tsx` provides us only R1. 

To get R2-4 as well, we could either:
O1. Create a separate utility function 'getCommitInfo' in 'github-helpers' (code below), OR
O2. Make the api.v3(...) call in the useful-not-found-page itself.

```ts
async function getCommitInfo(commitSha: string): Promise<Commit> { 
	// API v4 doesn't support it: https://github.community/t/what-is-the-corresponding-object-in-graphql-api-v4-for-patch-which-is-available-in-rest-api-v3/13590 
	const commit = await api.v3(`commits/${commitSha}`);
	return commit;
} 
```

What option would you suggest we go with?

Thanks!

Reply:
Thanks for pushing the change! Let me know if there is anything else to do before merging.